src/gluonts/model/rotbaum/_predictor.py

```diff
@@ -0,0 +1,184 @@
+# permissions and limitations under the License.
+
+# Standard library imports
+from enum import Enum
+from typing import Iterator, List, Optional, Tuple, Union
+
+# Third-party imports
+import numpy as np
+import pandas as pd
+from itertools import chain
+import concurrent.futures
+import logging
+
+# First-party imports
+from gluonts.core.component import validated
+from gluonts.dataset.common import Dataset
+from gluonts.model.forecast import Forecast, SampleForecast
+from gluonts.model.predictor import RepresentablePredictor
+from gluonts.support.pandas import forecast_start
+
+# Relative imports
+from ._preprocess import PreprocessOnlyLagFeatures
+from ._model import QRX
+
+
+class RotbaumForecast(Forecast):
+    """
+    Implements the quantile function in Forecast for TreePredictor,
+    as well as a new estimate_dists function for estimating a sampling of the
+    conditional distribution of the value of each of the steps in the
+    forecast horizon (independently).
+    """
+
+    @validated()
+    def __init__(
+        self,
+        models: List,
+        featurized_data: List,
+        start_date: pd.Timestamp,
+        freq,
+        prediction_length: int,
+    ):
+        self.models = models
+        self.featurized_data = featurized_data
+        self.start_date = start_date
+        self.freq = freq
+        self.prediction_length = prediction_length
+        self.item_id = None
+        self.lead_time = None
+
+    def quantile(self, q: float) -> np.array:
+        """
+        Returns np.array, where the i^th entry is the estimate of the q
+        quantile of the conditional distribution of the value of the i^th
+        step in the forecast horizon.
+        """
+        assert 0 <= q <= 1
+        return np.array(
+            list(
+                chain.from_iterable(
+                    model.predict(self.featurized_data, q)
+                    for model in self.models
+                )
+            )
+        )
+
+    def estimate_dists(self) -> np.array:
+        """
+        Returns np.array, where the i^th entry is an estimated sampling from
+        the conditional distribution of the value of the i^th step in the
+        forecast horizon.
+        """
+        return np.array(
+            list(
+                chain.from_iterable(
+                        model.estimate_dist(self.featurized_data)
+                        for model in self.models
+                )
+            )
+        )
+
+
+class TreePredictor(RepresentablePredictor):
+    """
+    A predictor that uses a QRX model for each of the steps in the forecast
+    horizon. (In other words, there's a total of prediction_length many
+    models being trained. In particular, this predictor does not learn a
+    multivariate distribution.) The list of these models is saved under
+    self.model_list.
+    """
+
+    @validated()
+    def __init__(
+        self,
+        context_length: Optional[int],
```

Comment:
replace with `context_length: Optional[int] = None,`.  `Optional` is only used when `None` is a potential value.  Also this will fix that `context_length` is not a required parameter.

Reply:
Done.